pkg/client/client.go

```diff
@@ -632,29 +649,70 @@ func (c *Client) GetAlertingRule(ctx context.Context, namespace, name string) (*
+	if err != nil {
+		return false, fmt.Errorf("converting Prometheus object to unstructured failed: %w", err)
+	}
+	unstructuredRequiredPrometheus := &unstructured.Unstructured{}
+	unstructuredRequiredPrometheus.SetUnstructuredContent(unstructuredRequiredPrometheusObject)
+
+	// Set the GVK for the unstructured object, since these are not inferred automatically, unlike their structured counterparts.
+	unstructuredRequiredPrometheus.SetGroupVersionKind(monv1.SchemeGroupVersion.WithKind(monv1.PrometheusesKind))
+
+	// Preserve the original behavior: always merge the metadata, never replace.
+	// Refer: https://github.com/openshift/cluster-monitoring-operator/pull/942.
+	prometheusGVR := unstructuredRequiredPrometheus.GroupVersionKind().GroupVersion().WithResource(monv1.PrometheusName)
+	unstructuredExistingPrometheus, err := c.dclient.
+		Resource(prometheusGVR).
+		Namespace(structuredRequiredPrometheus.GetNamespace()).
+		Get(ctx, structuredRequiredPrometheus.GetName(), metav1.GetOptions{})
 	if apierrors.IsNotFound(err) {
-		_, err := pclient.Create(ctx, p, metav1.CreateOptions{})
+		_, didUpdate, err := resourceapply.ApplyUnstructuredResourceImproved(
+			ctx,
+			c.dclient,
+			c.eventRecorder,
+			unstructuredRequiredPrometheus,
+			c.resourceCache,
+			prometheusGVR,
+			nil,
+			nil,
+		)
 		if err != nil {
-			return fmt.Errorf("creating Prometheus object failed: %w", err)
+			return didUpdate, fmt.Errorf("creating Prometheus object failed: %w", err)
 		}
-		return nil
+
+		return true, nil
 	}
 	if err != nil {
-		return fmt.Errorf("retrieving Prometheus object failed: %w", err)
+		return false, fmt.Errorf("retrieving Prometheus object failed: %w", err)
 	}
+	unstructuredRequiredPrometheusMetadataLabels := unstructuredRequiredPrometheus.GetLabels()
+	unstructuredRequiredPrometheusMetadataAnnotations := unstructuredRequiredPrometheus.GetAnnotations()
+	mergeMetadataLabels(unstructuredRequiredPrometheusMetadataLabels, unstructuredExistingPrometheus.GetLabels())
+	mergeMetadataAnnotations(unstructuredRequiredPrometheusMetadataAnnotations, unstructuredExistingPrometheus.GetAnnotations())
+	unstructuredRequiredPrometheus.SetLabels(unstructuredRequiredPrometheusMetadataLabels)
+	unstructuredRequiredPrometheus.SetAnnotations(unstructuredRequiredPrometheusMetadataAnnotations)
+	unstructuredRequiredPrometheus.SetResourceVersion(unstructuredExistingPrometheus.GetResourceVersion())
 
-	required := p.DeepCopy()
-	mergeMetadata(&required.ObjectMeta, existing.ObjectMeta)
-
-	required.ResourceVersion = existing.ResourceVersion
-	_, err = pclient.Update(ctx, required, metav1.UpdateOptions{})
+	_, didUpdate, err := resourceapply.ApplyUnstructuredResourceImproved(
+		ctx,
+		c.dclient,
+		c.eventRecorder,
+		unstructuredRequiredPrometheus,
+		c.resourceCache,
+		prometheusGVR,
+		nil,
+		nil,
+	)
 	if err != nil {
-		return fmt.Errorf("updating Prometheus object failed: %w", err)
+		return didUpdate, fmt.Errorf("updating Prometheus object failed: %w", err)
```

Comment:
I find the interface a bit confusing. The bool is ignored in all call sites except for tests. But the tests only test `shouldUpdate=true and err != nil`. Its not clear what that scenario actually means and why `CreateOrUpdatePrometheus` needs to add this to its signature.
We should at least have some clarifying comments.

Reply:
ACK, I believe the additional `bool` in the upstream signature is an artifact of a previous oversight which created scenarios where [an error could be returned with a true `didUpdate` value](https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/monitoring.go#L94). I'll drop the `bool` from the signature here since it contributes little, and also refactor upstream to do away with the anti-pattern.

Reply:
Added a comment to clarify the matrix.